File: src/algorithms/strings/transformation/string-compression/sources/string-compression.ts
// String Compression (Run-Length Encoding) — count consecutive repeated characters.
// Returns the compressed form "a2b1c5a3" only if shorter than the original; otherwise returns the original.
// Time: O(n)  Space: O(n) for the output buffer

export function stringCompression(text: string): string {
  if (text.length === 0) return text; // @step:initialize

  let compressed = ""; // @step:initialize
  let charIndex = 0; // @step:initialize

  while (charIndex < text.length) {
    const currentChar = text[charIndex] ?? ""; // @step:read-char
    let count = 0; // @step:read-char

    while (charIndex < text.length && text[charIndex] === currentChar) {
      count++; // @step:count
      charIndex++; // @step:count
    }

    compressed += currentChar; // @step:write-char
    compressed += String(count); // @step:write-char
  }

  return compressed.length < text.length ? compressed : text; // @step:complete
}

File: src/algorithms/strings/transformation/string-compression/sources/string-compression.go
// String Compression (Run-Length Encoding) — count consecutive repeated characters.
// Returns the compressed form "a2b1c5a3" only if shorter than the original; otherwise returns the original.
// Time: O(n)  Space: O(n) for the output buffer

package main

import (
	"strconv"
	"strings"
)

func stringCompression(text string) string {
	chars := []rune(text)
	if len(chars) == 0 { return text } // @step:initialize

	var compressedBuilder strings.Builder // @step:initialize
	charIndex := 0 // @step:initialize

	for charIndex < len(chars) {
		currentChar := chars[charIndex] // @step:read-char
		count := 0 // @step:read-char

		for charIndex < len(chars) && chars[charIndex] == currentChar {
			count++ // @step:count
			charIndex++ // @step:count
		}

		compressedBuilder.WriteRune(currentChar) // @step:write-char
		compressedBuilder.WriteString(strconv.Itoa(count)) // @step:write-char
	}

	compressed := compressedBuilder.String()
	if len([]rune(compressed)) < len(chars) { return compressed } // @step:complete
	return text // @step:complete
}
